File: python/version1/main.py
#!/usr/bin/python3
"""
    The Goal of this program is to generate a deck of cards.
    Ideally it will do so in "New Deck Format"
    It would be cool if I could shuffle the deck.
    It would be cool if I could deal some cards.
"""

import random


def main():
    """ Let's use this function for control flow, input and outputs"""
    print(f"Generating a new deck.")
    for x in range(100):
        card_deck = new_deck()
        for i in range(26):
            shuffle_deck(card_deck)
            card_deck = deal_cards(card_deck)


def new_deck():
    """ 
        This function should generate a new deck
        New Deck Order is: Joker1, Joker 2,
                            A->K Spades 
                            A->K Diamonds 
                            A->K Clubs 
                            A->K Hearts 
    """
    my_deck = [] 
    for suit in ("S", "D", "C", "H"):
        for value in ("A", "1", "2", "3", "4", "5", "6", "7", "8", "9", "J", "Q", "K"):
            my_deck.append(f'{value}{suit}')
    return my_deck


def shuffle_deck(card_deck):
    """
        This function should take the deck in whatever order it is
        and shuffle it.
    """
    random.shuffle(card_deck)
    

def deal_cards(my_deck):
    """
        This should take an integer as an input and deal that number
        of cards to each player.  Which means we should also accept a
        number of players as an input as well.
    """
    choice_one = my_deck[0]
    choice_two = my_deck[1]
    my_deck = my_deck[2:]
    print(f'{choice_one}, {choice_two}, {len(my_deck)}')
    return my_deck



if __name__ == '__main__':
    main()

File: cpp/version1/main.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <random>

using namespace std;

char suits[4] = {'S','D','C','H'};
char values[13] = {'A','1','2','3','4','5','6','7','8','9','J','Q','K'};

vector<string> new_deck() {
    vector<string> cards;
    for(int s = 0; s < 4; s++) {
        for(int v = 0; v < 13; v++) {
            cards.push_back(string({values[v], suits[s]}));
        }
    }
    return cards;
}

void shuffle_deck(vector<string> &card_deck) {
    auto rd = std::random_device {}; 
    auto rng = std::default_random_engine { rd() };
    shuffle(begin(card_deck), end(card_deck), rng);
}

void deal_cards(vector<string> & my_deck) {
    string choice_one = my_deck[0];
    string choice_two = my_deck[1];
    my_deck.erase(my_deck.begin(), my_deck.begin()+2);
    cout << choice_one << ", " << choice_two << ", " << my_deck.size() << "\n";
}

int main () {
    cout << "Generating a new deck.\n";
    for(int x = 0; x < 100; x++) {
        vector<string> cards = new_deck();
        for(int i = 0; i < 26; i++) {
            shuffle_deck(cards);
            deal_cards(cards);
        }
    }
}